src/zope/traversing/tests/test_publicationtraverse.py

```diff
@@ -134,13 +127,78 @@ def testTraverseRelativeURL(self):
 
     def testMissingSkin(self):
         ob = Content()
-        from zope.traversing.publicationtraverse import PublicationTraverser
         t = PublicationTraverser()
         request = TestRequest()
         self.assertRaises(
             NotFound, t.traversePath, request, ob, '/++skin++missingskin')
 
 
+    def test_traversePath_trailing_slash(self):
+        class Traverser(PublicationTraverser):
+            def __init__(self):
+                self.names = []
+
+            def traverseName(self, request, ob, name):
+                self.names.append(name)
+
+
+        t = Traverser()
+        t.traversePath(None, None, 'abc/def/')
+        self.assertEqual(t.names, ['abc', 'def'])
+
+        t = Traverser()
+        t.traversePath(None, None, 'abc/def///')
+
+        # Note that only *one* trailing slash is removed
+        self.assertEqual(t.names, ['abc', 'def', '', ''])
+
+
+    def test_traversePath_double_dots_cannot_remove(self):
+        class Traverser(PublicationTraverser):
+            def __init__(self):
+                self.names = []
+
+            def traverseName(self, request, ob, name):
+                self.names.append(name)
+
+
+        t = Traverser()
+        t.traversePath(None, None, '..')
+        self.assertEqual(t.names, ['..'])
+
+    def test_traverseRelativeURL_to_no_browser_publisher(self):
+        test = self
+        class Traverser(PublicationTraverser):
+            def traversePath(self, request, ob, path):
+                return ob
+
+
+        class Context(object):
+            called = False
+            def __conform__(self, iface):
+                self.called = True
+                test.assertEqual(iface, IBrowserPublisher)
+                return None
+
+        t = Traverser()
+        context = Context()
+        ob = t.traverseRelativeURL(None, context, None)
+        self.assertIs(ob, context)
+
+        self.assertTrue(context.called)
+
+class TestBeforeTraverseEvent(unittest.TestCase):
+
+    def test_interfaces(self):
+        from zope.traversing.interfaces import IBeforeTraverseEvent
+        from zope.traversing.interfaces import BeforeTraverseEvent
+        from zope.interface.verify import verifyObject
+
+        ob = BeforeTraverseEvent(self, self)
+        self.assertIs(self, ob.request)
+        self.assertIs(self, ob.object)
```

Comment:
Reusing `self` this way makes me sad.  Why not

```python
req, obj = object(), object()
ev = BeforeTraverseEvent(req, ob)
self.assertIs(ev.request, req)
self.assertIs(ev.object, obj)
verifyObject(IBeforeTraverseEvent, ev)
```

Reply:
Ok.